File: src/main/java/com/github/timo_reymann/ansible_vault_integration/settings/AnsibleVaultSettings.java
package com.github.timo_reymann.ansible_vault_integration.settings;

import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Settings for ansible vault execution
 */
@State(name = "com.github.timo_reymann.ansible_vault_integration.settings.AnsibleVaultSettings", storages = {@Storage("ansibleVaultSettings.xml")})
public class AnsibleVaultSettings implements PersistentStateComponent<AnsibleVaultSettingsState> {
    private AnsibleVaultSettingsState state;

    @Nullable
    @Override
    public AnsibleVaultSettingsState getState() {
        return state;
    }

    @Override
    public void loadState(@NotNull AnsibleVaultSettingsState state) {
        this.state = state;
        if (this.state.getVaultExecutable() == null || this.state.getVaultExecutable().isEmpty()) {
            this.setDefaultExecutable();
        }
    }

    public static AnsibleVaultSettings getInstance(Project project) {
        return ServiceManager.getService(project, AnsibleVaultSettings.class);
    }

    @Override
    public void noStateLoaded() {
        this.setDefaultExecutable();
    }

    public void setDefaultExecutable() {
        if (this.state == null) {
            this.state = new AnsibleVaultSettingsState();
        }

        this.state.setVaultExecutable(AnsibleVaultSettingsState.DEFAULT_VAULT_EXECUTABLE);
    }
}

File: src/main/kotlin/de/timo_reymann/ansible_vault_integration/action/settings/AnsibleVaultSettings.kt
package de.timo_reymann.ansible_vault_integration.action.settings

import com.intellij.openapi.components.*
import com.intellij.openapi.project.Project

/**
 * Settings for ansible vault execution
 */
@State(
    name = "com.github.timo_reymann.ansible_vault_integration.settings.AnsibleVaultSettings",
    storages = [Storage("ansibleVaultSettings.xml")]
)
open class AnsibleVaultSettings : PersistentStateComponent<AnsibleVaultSettings>, BaseState() {
    var vaultExecutable: String by nonNullString(DEFAULT_VAULT_EXECUTABLE)
    var vaultArguments: String by nonNullString()
    var timeout: Int by property(DEFAULT_TIMEOUT_SECONDS)

    override fun getState(): AnsibleVaultSettings = this

    override fun loadState(state: AnsibleVaultSettings) {
        copyFrom(state)
    }

    override fun noStateLoaded() {
        setDefaultExecutable()
    }

    private fun nonNullString(initialValue: String = "") = property(initialValue) { it == initialValue }

    private fun setDefaultExecutable() {
        vaultExecutable = DEFAULT_VAULT_EXECUTABLE
    }

    companion object {
        fun getInstance(project: Project) = project.service<AnsibleVaultSettings>()
        const val DEFAULT_TIMEOUT_SECONDS = 3
        const val DEFAULT_VAULT_EXECUTABLE = "/usr/bin/ansible-vault"
    }
}
